File: backend/src/controllers/draft-controller.js
import { z } from 'zod'
import { databasePool } from '../config/database.js'
import { errorResponse, successResponse } from '../utils/response.js'

const draftSchema = z.object({
  title: z.string().max(200).default(''),
  summary: z.string().max(2000).optional().default(''),
  content: z.string().max(100000).optional().default(''),
  categoryId: z.number().int().positive().nullable().optional(),
  riskInfo: z.record(z.any()).optional().default({})
})

function normalizeDraft(row) {
  return { id: row.id, title: row.title, summary: row.summary || '', content: row.content || '', categoryId: row.category_id, status: row.status, riskInfo: row.risk_info || {}, createdAt: row.created_at, updatedAt: row.updated_at }
}

export async function getDrafts(request, response, next) {
  try {
    const keyword = String(request.query.keyword || '').trim()
    const order = request.query.order === 'createdAt' ? 'created_at' : 'updated_at'
    const [rows] = await databasePool.query(`SELECT id,title,summary,content,category_id,status,risk_info,created_at,updated_at FROM drafts WHERE author_id = ? AND status = 'draft' AND (title LIKE ? OR summary LIKE ?) ORDER BY ${order} DESC`, [request.user.id, `%${keyword}%`, `%${keyword}%`])
    response.json(successResponse(rows.map(normalizeDraft)))
  } catch (error) { next(error) }
}

export async function createDraft(request, response, next) {
  try {
    const input = draftSchema.parse(request.body || {})
    const [result] = await databasePool.query('INSERT INTO drafts (title,summary,content,author_id,category_id,status,risk_info) VALUES (?,?,?,?,?,?,?)', [input.title, input.summary, input.content, request.user.id, input.categoryId || null, 'draft', JSON.stringify(input.riskInfo)])
    const [rows] = await databasePool.query('SELECT id,title,summary,content,category_id,status,risk_info,created_at,updated_at FROM drafts WHERE id = ? AND author_id = ?', [result.insertId, request.user.id])
    response.status(201).json(successResponse(normalizeDraft(rows[0]), '草稿已创建'))
  } catch (error) { next(error) }
}

export async function getDraftById(request, response, next) {
  try {
    const [rows] = await databasePool.query('SELECT id,title,summary,content,category_id,status,risk_info,created_at,updated_at FROM drafts WHERE id = ? AND author_id = ?', [request.params.id, request.user.id])
    if (!rows[0]) return response.status(404).json(errorResponse('草稿不存在或无权访问'))
    response.json(successResponse(normalizeDraft(rows[0])))
  } catch (error) { next(error) }
}

export async function updateDraft(request, response, next) {
  try {
    const input = draftSchema.parse(request.body || {})
    const [result] = await databasePool.query('UPDATE drafts SET title = ?, summary = ?, content = ?, category_id = ?, risk_info = ? WHERE id = ? AND author_id = ? AND status = \'draft\'', [input.title, input.summary, input.content, input.categoryId || null, JSON.stringify(input.riskInfo), request.params.id, request.user.id])
    if (!result.affectedRows) return response.status(404).json(errorResponse('草稿不存在或无权修改'))
    const [rows] = await databasePool.query('SELECT id,title,summary,content,category_id,status,risk_info,created_at,updated_at FROM drafts WHERE id = ? AND author_id = ?', [request.params.id, request.user.id])
    response.json(successResponse(normalizeDraft(rows[0]), '草稿已保存'))
  } catch (error) { next(error) }
}

export async function deleteDraft(request, response, next) {
  try {
    const [result] = await databasePool.query('DELETE FROM drafts WHERE id = ? AND author_id = ? AND status = \'draft\'', [request.params.id, request.user.id])
    if (!result.affectedRows) return response.status(404).json(errorResponse('草稿不存在或无权删除'))
    response.json(successResponse(null, '草稿已删除'))
  } catch (error) { next(error) }
}

export async function publishDraft(request, response, next) {
  const connection = await databasePool.getConnection()
  try {
    const [draftRows] = await connection.query('SELECT * FROM drafts WHERE id = ? AND author_id = ? AND status = \'draft\' FOR UPDATE', [request.params.id, request.user.id])
    const draft = draftRows[0]
    if (!draft) return response.status(404).json(errorResponse('草稿不存在或无权发布'))
    if (draft.title.trim().length < 2 || !draft.content.trim()) return response.status(400).json(errorResponse('发布前请填写标题和正文'))
    await connection.beginTransaction()
    const slug = `${draft.title.trim().toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '') || 'article'}-${Date.now()}`
    const [articleResult] = await connection.query('INSERT INTO articles (title,slug,summary,content,author_id,category_id,status,published_at) VALUES (?,?,?,?,?,?,?,NOW())', [draft.title, slug, draft.summary, draft.content, request.user.id, draft.category_id, 'published'])
    await connection.query('UPDATE drafts SET status = \'published\' WHERE id = ? AND author_id = ?', [draft.id, request.user.id])
    await connection.commit()
    response.status(201).json(successResponse({ articleId: articleResult.insertId }, '文章已发布'))
  } catch (error) { await connection.rollback(); next(error) } finally { connection.release() }
}

File: backend-java/src/main/java/com/myblog/draft/DraftController.java
package com.myblog.draft;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.myblog.common.ApiResponse;
import com.myblog.common.RowMapperUtil;
import com.myblog.security.UserPrincipal;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.sql.Statement;
import java.text.Normalizer;
import java.util.*;

@RestController
@RequestMapping("/api/drafts")
@PreAuthorize("hasAnyRole('author','admin')")
public class DraftController {
    private final JdbcTemplate jdbcTemplate;
    private final ObjectMapper objectMapper;

    public DraftController(JdbcTemplate jdbcTemplate, ObjectMapper objectMapper) {
        this.jdbcTemplate = jdbcTemplate;
        this.objectMapper = objectMapper;
    }

    @GetMapping
    public ApiResponse<List<Map<String, Object>>> list(
            @AuthenticationPrincipal UserPrincipal principal,
            @RequestParam(defaultValue = "") String keyword,
            @RequestParam(defaultValue = "updatedAt") String order
    ) {
        String search = "%" + keyword.trim() + "%";
        String orderBy = "createdAt".equals(order) ? "created_at" : "updated_at";
        return ApiResponse.ok(jdbcTemplate.queryForList("SELECT id, title, summary, content, author_id, category_id, status, risk_info, created_at, updated_at FROM drafts WHERE author_id = ? AND status = 'draft' AND (title LIKE ? OR summary LIKE ?) ORDER BY " + orderBy + " DESC", principal.userId(), search, search).stream().map(this::toDraft).toList());
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Map<String, Object>>> create(@AuthenticationPrincipal UserPrincipal principal, @RequestBody DraftRequest request) {
        KeyHolder keyHolder = new GeneratedKeyHolder();
        jdbcTemplate.update(connection -> {
            var statement = connection.prepareStatement(
                    "INSERT INTO drafts (title, summary, content, author_id, category_id, status, risk_info) VALUES (?, ?, ?, ?, ?, 'draft', ?)",
                    Statement.RETURN_GENERATED_KEYS
            );
            statement.setString(1, value(request.title()));
            statement.setString(2, value(request.summary()));
            statement.setString(3, value(request.content()));
            statement.setLong(4, principal.userId());
            statement.setObject(5, request.categoryId());
            statement.setString(6, riskInfo(request.riskInfo()));
            return statement;
        }, keyHolder);
        Number generatedId = keyHolder.getKey();
        if (generatedId == null) throw new IllegalStateException("草稿创建失败");
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok(findOwned(principal.userId(), generatedId.longValue()), "草稿已创建"));
    }

    @GetMapping("/{id}")
    public ApiResponse<Map<String, Object>> detail(@AuthenticationPrincipal UserPrincipal principal, @PathVariable long id) { return ApiResponse.ok(findOwned(principal.userId(), id)); }

    @PutMapping("/{id}")
    public ApiResponse<Map<String, Object>> update(@AuthenticationPrincipal UserPrincipal principal, @PathVariable long id, @RequestBody DraftRequest request) {
        int updatedRows = jdbcTemplate.update("UPDATE drafts SET title = ?, summary = ?, content = ?, category_id = ?, risk_info = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ? AND author_id = ? AND status = 'draft'", value(request.title()), value(request.summary()), value(request.content()), request.categoryId(), riskInfo(request.riskInfo()), id, principal.userId());
        if (updatedRows != 1) throw new NoSuchElementException("草稿不存在或无权修改");
        return ApiResponse.ok(findOwned(principal.userId(), id), "草稿已保存");
    }

    @DeleteMapping("/{id}")
    public ApiResponse<Void> delete(@AuthenticationPrincipal UserPrincipal principal, @PathVariable long id) {
        int deletedRows = jdbcTemplate.update("DELETE FROM drafts WHERE id = ? AND author_id = ? AND status = 'draft'", id, principal.userId());
        if (deletedRows != 1) throw new NoSuchElementException("草稿不存在或无权删除");
        return ApiResponse.ok(null, "草稿已删除");
    }

    @PostMapping("/{id}/publish")
    @Transactional
    public ResponseEntity<ApiResponse<Map<String, Object>>> publish(@AuthenticationPrincipal UserPrincipal principal, @PathVariable long id) {
        Map<String, Object> draft = findOwnedDraft(principal.userId(), id);
        String title = value(draft.get("title")).trim();
        String content = value(draft.get("content")).trim();
        if (title.length() < 2) throw new IllegalArgumentException("标题至少需要 2 个字");
        if (content.isEmpty()) throw new IllegalArgumentException("正文不能为空");

        int publishedRows = jdbcTemplate.update("UPDATE drafts SET status = 'published', updated_at = CURRENT_TIMESTAMP WHERE id = ? AND author_id = ? AND status = 'draft'", id, principal.userId());
        if (publishedRows != 1) throw new NoSuchElementException("草稿不存在或无权发布");

        String slug = makeSlug(title + "-" + System.currentTimeMillis());
        KeyHolder keyHolder = new GeneratedKeyHolder();
        jdbcTemplate.update(connection -> {
            var statement = connection.prepareStatement("INSERT INTO articles (title, slug, summary, content, author_id, category_id, status, published_at) VALUES (?, ?, ?, ?, ?, ?, 'published', CURRENT_TIMESTAMP)", Statement.RETURN_GENERATED_KEYS);
            statement.setString(1, title);
            statement.setString(2, slug);
            statement.setString(3, value(draft.get("summary")));
            statement.setString(4, content);
            statement.setLong(5, principal.userId());
            statement.setObject(6, draft.get("categoryId"));
            return statement;
        }, keyHolder);
        Number generatedId = keyHolder.getKey();
        if (generatedId == null) throw new IllegalStateException("文章发布失败");

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("articleId", generatedId.longValue());
        result.put("slug", slug);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok(result, "文章已发布"));
    }

    private Map<String, Object> findOwned(long userId, long id) {
        List<Map<String, Object>> rows = jdbcTemplate.queryForList("SELECT id, title, summary, content, author_id, category_id, status, risk_info, created_at, updated_at FROM drafts WHERE id = ? AND author_id = ?", id, userId);
        if (rows.isEmpty()) throw new NoSuchElementException("草稿不存在或无权访问");
        return toDraft(rows.getFirst());
    }

    private Map<String, Object> findOwnedDraft(long userId, long id) {
        List<Map<String, Object>> rows = jdbcTemplate.queryForList("SELECT id, title, summary, content, author_id, category_id, status, risk_info, created_at, updated_at FROM drafts WHERE id = ? AND author_id = ? AND status = 'draft'", id, userId);
        if (rows.isEmpty()) throw new NoSuchElementException("草稿不存在或无权发布");
        return toDraft(rows.getFirst());
    }

    private Map<String, Object> toDraft(Map<String, Object> row) {
        Map<String, Object> result = RowMapperUtil.camel(row);
        Object rawRiskInfo = result.get("riskInfo");
        if (rawRiskInfo instanceof String riskInfo) {
            try { result.put("riskInfo", objectMapper.readTree(riskInfo)); }
            catch (JsonProcessingException ignored) { result.put("riskInfo", objectMapper.createObjectNode()); }
        }
        if (rawRiskInfo == null) result.put("riskInfo", objectMapper.createObjectNode());
        return result;
    }

    private String value(String value) { return value == null ? "" : value; }
    private String value(Object value) { return value == null ? "" : String.valueOf(value); }
    private String riskInfo(JsonNode value) { return value == null || value.isNull() ? "{}" : value.toString(); }
    private String makeSlug(String value) { return Normalizer.normalize(value.toLowerCase(Locale.ROOT), Normalizer.Form.NFKD).replaceAll("[^a-z0-9]+", "-").replaceAll("^-|-$", ""); }

    public record DraftRequest(String title, String summary, String content, Integer categoryId, JsonNode riskInfo) {}
}
